python/ql/src/experimental/semmle/python/frameworks/Django.qll

```diff
@@ -7,4 +7,260 @@ private import experimental.dataflow.DataFlow
+      private DataFlow::Node cursor(DataFlow::TypeTracker t) {
+        t.start() and
+        result.asCfgNode().(CallNode).getFunction() = classCursor().asCfgNode()
+        or
+        exists(DataFlow::TypeTracker t2 | result = cursor(t2).track(t2, t))
+      }
+
+      /** Gets a reference to an instance of `django.db.connection.cursor`. */
+      DataFlow::Node cursor() { result = cursor(DataFlow::TypeTracker::end()) }
+
+      /** Gets a reference to the `django.db.connection.cursor.execute` function. */
+      private DataFlow::Node execute(DataFlow::TypeTracker t) {
+        t.startInAttr("execute") and
+        result = cursor()
+        or
+        exists(DataFlow::TypeTracker t2 | result = execute(t2).track(t2, t))
+      }
+
+      /** Gets a reference to the `django.db.connection.cursor.execute` function. */
+      DataFlow::Node execute() { result = execute(DataFlow::TypeTracker::end()) }
+
+      /** Gets a reference to the `django.db.models` module. */
+      private DataFlow::Node models(DataFlow::TypeTracker t) {
+        t.start() and
+        result = DataFlow::importNode("django.db.models")
+        or
+        t.startInAttr("models") and
+        result = django()
+        or
+        exists(DataFlow::TypeTracker t2 | result = models(t2).track(t2, t))
+      }
+
+      /** Gets a reference to the `django.db.models` module. */
+      DataFlow::Node models() { result = models(DataFlow::TypeTracker::end()) }
+
+      /** Provides models for the `django.db.models` module. */
+      module models {
+        /** Gets a reference to the `django.db.models.Model` class. */
+        private DataFlow::Node classModel(DataFlow::TypeTracker t) {
+          t.start() and
+          result = DataFlow::importNode("django.db.models.Model")
+          or
+          t.startInAttr("Model") and
+          result = models()
+          or
+          exists(DataFlow::TypeTracker t2 | result = classModel(t2).track(t2, t))
+        }
+
+        /** Gets a reference to the `django.db.models.Model` class. */
+        DataFlow::Node classModel() { result = classModel(DataFlow::TypeTracker::end()) }
+
+        /** Gets a definition of a subclass the `django.db.models.Model` class. */
+        class ClassModelSubclassDef extends ControlFlowNode {
+          string name;
+
+          ClassModelSubclassDef() {
+            exists(ClassExpr ce |
+              this.getNode() = ce and
+              ce.getABase() = classModel().asExpr() and
+              ce.getName() = name
+            )
+          }
+
+          string getName() { result = name }
+        }
+
+        /**
+         * A reference to a class that is a subclass of the `django.db.models.Model` class.
+         * This is quite an approximation, since it simply matches identifiers.
+         */
+        class ClassModelSubclass extends DataFlow::CfgNode {
+          override NameNode node;
+
+          ClassModelSubclass() { node.getId() = any(ClassModelSubclassDef cd).getName() }
+        }
```

Comment:
Why don't you use a type-tracker for this?

Reply:
I did not see how to do that, where do I make it start?

Reply:
Oh, I make it start where it is defined now and then track..